### ISSUE TITLE
Gradient accumulation in recurrent architectures

### ISSUE THREAD

**Issue by jsuarez5341** [pytorch/pytorch#1407]
Suppose I have some recurrent network. One important consideration in analysis of new architectures is the gradient flow. As far as I am aware, there is currently no way to access the gradients wrt the weights at each time step, as they are accumulated internally, even when retain_variables is specified in backward().

This seems like a rather essential feature for research. Am I missing the documentation, or is there currently no way to specify that backward() should not accumulate gradients?

**Comment by soumith** [comment]
if you are using RNNCell / LSTMCell, then you can register a hook on the weights to access individual weights.
If you are using nn.RNN / nn.LSTM, then there is no way on the GPU because CuDNN does the whole operation in a single CUDA kernel, i.e. there is no individual ops.

**Comment by jsuarez5341** [comment]
I was afraid you might say that. Also, I'm not trying to access individual weights (that I can do), but gradients of those weights at different timesteps. Assume it's a cell (though it's one I wrote in this case). I'm still not quite sure how I would do this from the register_hook docs: "The hook will be called every time a gradient with respect to the variable is computed." ... So I suppose I could define a closure for each separate variable that adds the gradient values to a list? If this sounds right, I can do this for now. I'm wondering for long term usability if there's any reason not to have an option to, say, add a keyword option accumulate_gradients=False to backward?

**Comment by soumith** [comment]
> a closure for each separate variable that adds the gradient values to a list?
Yes, what you've proposed sounds about right.

>  I'm wondering for long term usability if there's any reason not to have an option to, say, add a keyword option accumulate_gradients=False to backward?

It's not obvious what one wants to do there. Do we just overwrite gradients? or keep a list of gradients?
The semantics are hard and non-obvious, and the overall API becomes much harder to use.
However, in your benefit, we introduced an API on `master` (not yet on any binary releases) that will get you gradient wrt an input, independent of using `.grad`, i.e. it doesn't accumulate gradients: http://pytorch.org/docs/autograd.html#torch.autograd.grad
It's more powerful, but also much more verbose to use.

**Comment by apaszke** [comment]
A simple hack that would allow you to get what you want:
```python
def cell(input, hidden, _weight):
    weight = _weight.view(_weight.size()) # or any cheap op in here (just to create a graph node)
    weight.register_hook(lambda grad: print(grad)) # this will be called only for a single time step
    return weight.mm(input) + weight.mm(hidden) # your code goes here
```

You could also use `torch.autograd.grad`, but it would be slow to compute the backward `num_timesteps` times. I'm not sure how could we add support for such things in autograd - backend treats a Variable as a single value used to compute the output, and I think what you want to compute isn't well defined mathematically (it's not `dL/dW`).

**Comment by ROZBEH** [comment]
@jsuarez5341 Were you able to solve the issue? I want to do exactly the same thing and calculate the gradients with respect to weights at each time step.
